kad/src/query.rs

```diff
@@ -254,18 +261,23 @@ where
             // `message` contains the reason why the current future was woken up.
             let closer_peers = match message {
                 Ok(msg) => msg,
-                Err(_) => {
+                Err(err) => {
+                    trace!(target: "libp2p-kad", "RPC query failed for {:?}: {:?}", remote_id, err);
                     state.failed_to_contact.insert(remote_id);
                     return Ok(future::Loop::Continue(state));
                 }
             };
 
-            // Update `state.result` with the fact that we received a valid message from a node.
+            // Inserting the node we received a response from into `state.result`.
+            // The code is non-trivial because `state.result` is ordered by distance and is limited
+            // by `num_results` elements.
             if let Some(insert_pos) = state.result.iter().position(|e| {
                 e.distance_with(&searched_key) >= remote_id.distance_with(&searched_key)
             }) {
                 if state.result[insert_pos] != remote_id {
-                    state.result.pop();
+                    if state.result.len() >= num_results {
```

Comment:
Bugfix here: the reason why we pop an entry is to not go over `num_results` ; this was completely misimplemented